### PR TITLE
DAOS-6317 pool: support for admin-defined domain in pmap

### PR DESCRIPTION
**Pull request by johannlombardi**
Some changes are required to the pool map code to support variable
number of fault domain.

This patch implements the following changes:
- take a level (starting at 0 from the leaf = target) for the
container property.
- target and rank levels are hardcoded to respectively 0 and 1.
- 253 levels above nodes are then available to the user. The
control plane will maintain the association level ID to
name reported in the yaml file.
- change the component ID (= level) to 8 bits (instead of 16) for
a total of 256 which is way more than enough.

Signed-off-by: Johann Lombardi <johann.lombardi@intel.com>

- Fix some checks that still expected the previous ordering of
  fault domain levels.
- Added logging (and raised the level of some logging) for fatal
  errors in the pool map creation process.
- Renamed misleading component types. What was previously "NODE"
  is now "RANK" (i.e. referring to the specific engine). The
  level above that (previously "RACK") is "NODE", referring to a
  specific machine which may run multiple engines. Everything
  above that is arbitrary.

Signed-off-by: Kris Jacque <kristin.jacque@intel.com>

### CODE REVIEW

**Review by kjacque — DISMISSED** [review]

File: src/include/daos_prop.h [hunk 1]
```diff
@@ -271,10 +271,14 @@ enum {
 	DAOS_PROP_CO_REDUN_RF4,
 };
 
-/** container redundancy level */
+/**
+ * Level of fault-domain to use for object allocation
+ * node is hardcoded to 1, [2-254] are defined by the admin
```
Comment (NiuYawei) [inline src/include/daos_prop.h:276]: s/node/rank

**Review by NiuYawei — DISMISSED** [review]

File: src/include/daos/pool_map.h [hunk 2]
```diff
@@ -23,14 +23,18 @@
...
-	PO_COMP_TP_NODE		= 40,
-	PO_COMP_TP_TARGET	= 50,
-	/* TODO: more types */
+	PO_COMP_TP_TARGET	= 0, /** reserved, hard-coded */
+	PO_COMP_TP_RANK		= 1, /** reserved, hard-coded */
+	PO_COMP_TP_MIN		= 2, /** first user-defined domain */
+	PO_COMP_TP_NODE		= 2, /** for test only */
+	PO_COMP_TP_MAX		= 254, /** last user-defined domain */
+	PO_COMP_TP_ROOT		= 255,
+	PO_COMP_TP_END		= 256,
```
Comment (wangdi1) [inline src/include/daos/pool_map.h:37]: Hmm, it seems co_type is changed now, so you might need update POOL_MAP_VERSION to POOL_MAP_VER_2 ?

Comment (kjacque) [reply src/include/daos/pool_map.h:38]: Good catch, updated.

File: src/include/daos_prop.h (hunk 1 above)
Comment (kjacque) [reply src/include/daos_prop.h]: Fixed

**Review by NiuYawei — APPROVED** [review]

**Review by wangdi1 — APPROVED** [review]

**Pull request merged**